packages/fuels-core/src/abi_decoder.rs

```diff
@@ -1,252 +1,288 @@
-                    token: Token::Array(tokens),
-                    new_offset,
-                };
-
-                Ok(result)
-            }
-            ParamType::Struct(props) => {
-                let mut tokens = vec![];
-
-                let mut new_offset = offset;
-                for prop in props {
-                    let res = self.decode_param(prop, data, new_offset)?;
-                    new_offset = res.new_offset;
-                    tokens.push(res.token);
-                }
-
-                let result = DecodeResult {
-                    token: Token::Struct(tokens),
-                    new_offset,
-                };
-
-                Ok(result)
-            }
-            ParamType::Enum(variations) => {
-                let discriminant = peek_word(data, offset).unwrap();
-
-                let discriminant = u32::from_be_bytes(discriminant[4..8].try_into().unwrap());
-
-                // Offset + 8 because of the discriminant that we just peeked
-                let res = self.decode_param(
-                    variations.get(discriminant as usize).unwrap(),
-                    data,
-                    offset + 8,
-                )?;
-
-                let result = DecodeResult {
-                    token: Token::Enum(Box::new((discriminant as u8, res.token))),
-                    new_offset: res.new_offset,
-                };
-
-                Ok(result)
-            }
-            ParamType::Tuple(types) => {
-                let mut tokens = vec![];
-                let mut new_offset = offset;
-
-                for t in types {
-                    let res = self.decode_param(t, data, new_offset)?;
-                    new_offset = res.new_offset;
-                    tokens.push(res.token);
-                }
-
-                let result = DecodeResult {
-                    token: Token::Tuple(tokens),
-                    new_offset,
-                };
-
-                Ok(result)
-            }
+        let mut tokens = vec![];
+        let mut bytes_read = 0;
+
+        for _ in 0..*length {
+            let res = Self::decode_param(t, &data[bytes_read..])?;
+            bytes_read += res.bytes_read;
+            tokens.push(res.token);
         }
+
+        let result = DecodeResult {
+            token: Token::Array(tokens),
+            bytes_read,
+        };
+
+        Ok(result)
     }
-}
 
-impl Default for ABIDecoder {
-    fn default() -> Self {
-        Self::new()
+    fn decode_string(data: &[u8], length: &usize) -> Result<DecodeResult, CodecError> {
+        let encoded_str = peek(data, *length)?;
+
+        let decoded = str::from_utf8(encoded_str)?;
+
+        let result = DecodeResult {
+            token: Token::String(decoded.to_string()),
+            bytes_read: padded_len(encoded_str),
+        };
+
+        Ok(result)
+    }
+
+    fn decode_b256(data: &[u8]) -> Result<DecodeResult, CodecError> {
+        let b256: Bits256 = peek(data, 32)?.try_into().unwrap();
```

Comment:
For the sake of consistency and proper error handling, what do you think about turning this `try_into().unwrap()` into a proper `try_into()?` converting the error to `CodecError`? 

I can imagine this being very annoying to debug if there's something fishy in `data` and we hit this `unwrap()`.

Reply:
Sure! This was just copied&pasted from the original impl.

Reply:
Yeah, that makes sense, then let's do `expect("something something. This should never happen")` so that at least we know where to look at if the underlying impl ever breaks.

Reply:
There are no unwraps left in the production code in this file. Can you see if you're ok with the solution?